client/dfclient/src/menu_state.cpp

```diff
@@ -14,40 +18,119 @@
+{
+	nc::SceneNode &rootNode = nc::theApplication().rootNode();
+	auto res = nc::theApplication().appConfiguration().resolution;
+	auto text = new ncine::TextNode(&rootNode, _resources.fonts["comic"].get());
+	text->setString(message.c_str());
+	text->setPosition(res.x * 0.5f, res.y * 0.30f);
+	text->setScale(0.5f);
+	text->setColor(0, 0, 0, 255);
+	_resources._tweens.push_back(CreateTextTween(text));
+}
+
 MenuState::MenuState(Resources& r) : _resources(r) {
 	nc::SceneNode &rootNode = nc::theApplication().rootNode();
 	auto res = nc::theApplication().appConfiguration().resolution;
 	logoSprite = nctl::makeUnique<nc::Sprite>(&rootNode, _resources.textures["deadfish.png"].get(), res.x*0.5f, res.y*0.6f);
 	nc::theApplication().gfxDevice().setClearColor(bgColor);
 	if (gameData.gameInProgress) {
 		std::cout << "menu game in progress\n";
-		auto text = new ncine::TextNode(&rootNode, _resources.fonts["comic"].get());
-		text->setString("game already in progress");
-		text->setPosition(res.x * 0.5f, res.y * 0.75f);
-		text->setScale(2.0f);
-		text->setColor(0, 0, 0, 255);
-		_resources._tweens.push_back(CreateTextTween(text));
+		ShowMessage("game already in progress");
 		gameData.gameInProgress = false;
 	}
+	boost::property_tree::ptree pt;
+	boost::property_tree::ini_parser::read_ini("deadfish.ini", pt);
+	try {
+		this->matchmakerAddress = pt.get<std::string>("default.matchmaker_address", "");
+		this->matchmakerPort = pt.get<std::string>("default.matchmaker_port", "");
+		this->directAddress = pt.get<std::string>("default.direct_address", "");
+		this->directPort = pt.get<std::string>("default.direct_port", "");
+		std::string mode = pt.get<std::string>("default.client_mode");
+		if (mode == "matchmaker")
+			this->clientMode = ClientMode::Matchmaker;
+		else if (mode == "direct")
+			this->clientMode = ClientMode::DirectConnect;
+		else {
+			throw std::runtime_error(("unknown client mode: " + mode).c_str());
+		}
+	} catch (const std::exception& e) {
+		std::cout << "failed to read and parse deadfish.ini: " << e.what() << "\n";
+		exit(1);
+	}
+	if (this->clientMode == ClientMode::DirectConnect) {
+		auto addr = this->directAddress + ":" + this->directPort;
+		strcpy(this->buf0, addr.c_str());
+	}
 }
 
-bool MenuState::TryConnect() {
+void MenuState::TryConnect() {
 	gameData.serverAddress = "ws://" + gameData.serverAddress;
 	std::cout << "server " << gameData.serverAddress << ", my nickname " << gameData.myNickname << "\n";
 	gameData.socket = CreateWebSocket();
 	int ret = gameData.socket->Connect(gameData.serverAddress);
 	if (ret < 0) {
 		std::cout << "socket->Connect failed " << ret << "\n";
-		// TODO: some ui error handling
-		return false;
+		this->ShowMessage("connecting to server failed");
+	}
+}
+
+void MenuState::ProcessMatchmakerData(std::string data) {
+	boost::property_tree::ptree tree;
+	std::stringstream ss;
+	ss << data;
+	boost::property_tree::read_json(ss, tree);
+
+	auto status = tree.get("status", "");
```

Comment:
Is the server-matchmaker protocol documented somewhere?

Reply:
It is now in the README